dietpi/dietpi-services

```diff
@@ -40,644 +48,602 @@
+	Save_Process_Tool(){
 
-		# status
-		elif [[ $mode == 'status' ]]; then
+		local index=$1
 
-			for i in ${!aSERVICE_NAME[@]}
-			do
+		# Backwards cap for init.d, create systemd wrapper
+		if [[ ${aFP_SERVICE[$index]} == '/etc/init.d/'* ]]; then
 
-				# Apply
-				if (( ${aSERVICE_AVAILABLE[$i]} )); then
+			aFP_SERVICE[$index]="/etc/systemd/system/${aSERVICE_NAME[$index]}.service"
+			cat << _EOF_ > "${aFP_SERVICE[$index]}"
+[Unit]
+Description=${aSERVICE_NAME[$index]} sysvinit wrapper (DietPi)
 
-					local status="${aSERVICE_NAME[$i]}\t$(systemctl status ${aSERVICE_NAME[$i]} | grep -m1 'Active' | cut -c12-)"
-					if [[ $status =~ 'failed' ]]; then
+[Service]
+Type=forking
+ExecStart=/etc/init.d/${aSERVICE_NAME[$index]} start
+ExecStop=/etc/init.d/${aSERVICE_NAME[$index]} stop
 
-						G_DIETPI-NOTIFY 1 "$status"
-						systemctl -l --no-pager status ${aSERVICE_NAME[$i]}
+[Install]
+WantedBy=multi-user.target
+_EOF_
+			G_WHIP_MSG "[ INFO ] DietPi has created a systemd service wrapper for /etc/init.d/${aSERVICE_NAME[$index]}. This is required to allow support for applying process tool options.\n
+If you experience any issues with the wrapper, simply remove the file:\n - rm /etc/systemd/system/${aSERVICE_NAME[$index]}.service\n - Then restart the service."
 
-					elif [[ $status =~ 'inactive' ]]; then
+		fi
 
-						G_DIETPI-NOTIFY 2 "$status"
+		# Always create drop-in configs in: /etc/systemd/system/
+		mkdir -p "/etc/systemd/system/${aSERVICE_NAME[$index]}.service.d"
+		cat << _EOF_ > "/etc/systemd/system/${aSERVICE_NAME[$index]}.service.d/$FP_PROCESS_TOOL_CONF"
+# WARNING: Do not manually edit this file, use "dietpi-services" to adjust values!
+[Service]
+Nice=${aCPU_NICE[$index]}
+CPUAffinity=${aCPU_AFFINITY[$index]}
+CPUSchedulingPolicy=${aCPU_SCHEDULE_POLICY[$index]}
+CPUSchedulingPriority=${aCPU_SCHEDULE_PRIORITY[$index]}
+IOSchedulingClass=${aIO_SCHEDULE_POLICY[$index]}
+IOSchedulingPriority=${aIO_PRIORITY[$index]}
+_EOF_
+		G_RUN_CMD systemctl daemon-reload
+		aSERVICE_RESTART_REQUIRED[$index]=1
 
-					else
+	}
 
-						G_DIETPI-NOTIFY 0 "$status"
+	# $1=index
+	Load_Process_Tool(){
 
-					fi
+		local index=$1
 
-				fi
+		# Defaults
+		aCPU_NICE[$index]=0
+		aCPU_AFFINITY[$index]="0-$(( $G_HW_CPU_CORES - 1 ))"
+		aCPU_SCHEDULE_POLICY[$index]='other'
+		aCPU_SCHEDULE_PRIORITY[$index]=0
+		aIO_SCHEDULE_POLICY[$index]='best-effort'
+		aIO_PRIORITY[$index]='4' # best-effort default == (cpu_nice + 20) / 5: https://manpages.debian.org/stretch/util-linux/ionice.1.en.html
+
+		local fp="/etc/systemd/system/${aSERVICE_NAME[$index]}.service.d/$FP_PROCESS_TOOL_CONF"
+		[[ -f $fp ]] || return
+
+		# Source values from config file
+		# - [Service] line throws an effectless error that we can simply hide.
+		# - All values are single word strings, so bash assigns them correctly.
+		local Nice CPUAffinity CPUSchedulingPolicy CPUSchedulingPriority IOSchedulingClass IOSchedulingPriority
+		. $fp &> /dev/null
```

Comment:
@MichaIng 

Nice, I did think of adding ```#!/bin/bash``` to our drop in service, but interesting its not needed :+1:

Reply:
@Fourdee 
Hehe nope for sourcing it's not required. Actually it is even ignored. When sourcing scripts, they are always executed with the originating shell, of course, since no new shell is initiated.

Only downside is that custom user changes "might" be a danger. Although all systemd unit syntax is var=value style, so the worst thing is that we source some inerte variables and/or get some more silenced syntax errors.